grpc/transport-auth.go

```diff
@@ -0,0 +1,70 @@
+package grpc
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// MultiNameAuthenticator does a thing
+type MultiNameAuthenticator struct {
+	names map[string]*tls.Config
+}
+
+// NewMultiNameTLS also does a thing
+func NewMultiNameTLS(names map[string]*tls.Config) credentials.TransportAuthenticator {
+	return &MultiNameAuthenticator{names}
+}
+
+func (mna *MultiNameAuthenticator) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	var errChannel chan error
+	if timeout != 0 {
+		errChannel = make(chan error, 2)
+		time.AfterFunc(timeout, func() {
+			errChannel <- errors.New("boulder/grpc: Dial timed out")
+		})
+	}
+	config, present := mna.names[addr]
+	if !present {
+		return nil, nil, fmt.Errorf("boulder/grpc: Unexpected name, no TLS configuration present for \"%s\"", addr)
+	}
+	conn := tls.Client(rawConn, config)
+	var err error
+	if timeout == 0 {
+		err = conn.Handshake()
+	} else {
+		go func() {
+			errChannel <- conn.Handshake()
```

Comment:
Also, I think the "Dial timed out" error is inaccurate. The fact that we already have a `net.Conn` means the dial has happened. So it's the handshake that timed out.

And maybe we can use `conn.SetDeadline(..)` instead of doing our own timeout?